master/internal/command/command.go

```diff
@@ -19,124 +20,127 @@ import (
+	// Locking in: toNTSC, startCmd, onExit, deleteWorkspace.
 
-		GenericCommandSpec: *spec,
+	command struct {
+		mu sync.Mutex
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  jobType,
-		jobID:    jobID,
+		db *db.PgDB
+		rm rm.ResourceManager
 
-		logCtx: logger.Context{
-			"job-id":    jobID,
-			"task-id":   taskID,
-			"task-type": taskType,
-		},
+		tasks.GenericCommandSpec
 
-		contextDirectory: contextDirectory,
+		registeredTime time.Time
+		taskID         model.TaskID
+		taskType       model.TaskType
+		jobType        model.JobType
+		jobID          model.JobID
+		allocationID   model.AllocationID
+		lastState      task.AllocationState
+		exitStatus     *task.AllocationExited
+		restored       bool
+
+		contextDirectory []byte // Don't rely on this being set outsides of PreStart non restore case.
+
+		logCtx logger.Context
+		syslog *logrus.Entry
 	}
 
-	a, _ := ctx.ActorOf(cmd.taskID, cmd)
-	summaryFut := ctx.Ask(a, getSummary{})
-	if err := summaryFut.Error(); err != nil {
-		return errors.Wrap(err, "failed to create generic command")
+	// CreateGeneric is a request to the commandService to create a generic command.
+	CreateGeneric struct {
+		ContextDirectory []byte
+		Spec             *tasks.GenericCommandSpec
 	}
-	// Sync with the actor, but we don't really need the summary. actor.Ping works too,
-	// but this makes sure it can form some sort of useful response (ping doesn't actually
-	// hit the receive block).
-	summaryFut.Get()
-	return nil
-}
+)
 
-func commandFromSnapshot(
-	ctx *actor.Context,
-	db *db.PgDB,
-	rm rm.ResourceManager,
-	snapshot *CommandSnapshot,
-) *command {
-	taskID := snapshot.TaskID
-	taskType := snapshot.Task.TaskType
-	jobID := snapshot.Task.Job.JobID
-	cmd := &command{
-		db:             db,
-		rm:             rm,
-		registeredTime: snapshot.RegisteredTime,
+// InitCommandService initializes the global command service singleton & upon start,
+// restores all terminated commands whose end time isn't set.
+// InitCommandService locks ahead of restoreAllCommands -> remakeAllCommands -> commandFromSnapshot.
+func InitCommandService(ctx context.Context, pgDB *db.PgDB, rm rm.ResourceManager) {
+	if DefaultCmdService != nil {
+		logrus.Warn(
+			"detected re-initialization of command service that should never occur outside of tests",
+		)
+	}
 
-		GenericCommandSpec: snapshot.GenericCommandSpec,
+	DefaultCmdService = &commandService{
+		db:       pgDB,
+		rm:       rm,
+		commands: make(map[model.TaskID]*command),
+		syslog:   logrus.WithField("component", "command-service"),
+	}
 
-		taskID:   taskID,
-		taskType: taskType,
-		jobType:  snapshot.Task.Job.JobType,
-		jobID:    jobID,
+	DefaultCmdService.mu.Lock()
+	defer DefaultCmdService.mu.Unlock()
 
-		logCtx: logger.Context{
-			"job-id":    jobID,
-			"task-id":   taskID,
-			"task-type": taskType,
-		},
+	if err := DefaultCmdService.restoreAllCommands(ctx, pgDB, rm); err != nil {
+		DefaultCmdService.syslog.Error(err)
+	}
```

Comment:
i'd make a `newCommandService` and have it do this, or just make `restoreAllCommands` do the locking.

Reply:
Since I think we want locks in public functions, then I think we _shouldn't_ lock in `restoreAllCommands`.